src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/EnvelopeReprocessService.java

```diff
@@ -28,5 +43,50 @@ public void reprocessEnvelope(String envelopeId) {
             .orElseThrow(
                 () -> new EnvelopeNotFoundException("Envelope with id " + envelopeId + " not found")
             );
+
+        validateEnvelopeState(envelope);
+
+        createEvent(envelope);
+
+        envelope.setStatus(Status.UPLOADED);
+        envelopeRepository.save(envelope);
+    }
+
+    private void createEvent(Envelope envelope) {
+        ProcessEvent event = new ProcessEvent(
+            envelope.getContainer(),
+            envelope.getZipFileName(),
+            MANUAL_RETRIGGER_PROCESSING
+        );
+        event.setReason("Moved to UPLOADED status to reprocess the envelope");
+        processEventRepository.save(event);
+    }
+
+    private void validateEnvelopeState(Envelope envelope) {
+        if (envelope.getCcdId() != null) {
+            throw new EnvelopeProcessedInCcdException(
+                "Envelope with id " + envelope.getId() + " has already been processed in CCD"
+            );
+        }
+
+        if (envelope.getStatus() != COMPLETED && !isStale(envelope)) {
+            throw new EnvelopeNotCompletedOrStaleException(
+                "Envelope with id " + envelope.getId() + " is not completed or stale"
+            );
+        }
+    }
+
+    private boolean isStale(Envelope envelope) {
+        if (envelope.getStatus() != Status.NOTIFICATION_SENT) {
+            return false;
+        }
+
+        Instant lastEventTimeStamp = processEventRepository
+            .findByZipFileName(envelope.getZipFileName())
+            .stream()
+            .map(ProcessEvent::getCreatedAt)
+            .max(naturalOrder())
+            .get(); // no events for the envelope is normally impossible
+        return between(lastEventTimeStamp, now()).toMillis() > notificationTimeoutMs;
```

Comment:
`                getCreatedAt
                  .plus(notificationTimeoutHr)
                   .isBefore(now())`

this can be done?

Reply:
Fixed